spec/support/helpers.rb

```diff
@@ -7,6 +7,17 @@ def time_anchor
   end
 
   def time_anchor_unix
-    BigDecimal.new('1450150205.123')
+    BigDecimal('1450150205.123')
+  end
+
+  # Creates a Loga module mock with corresponding logger mock.
+  # Both mocks are verifiable.
+  #
+  # logger, loga = loga_mock
+  def loga_mock
+    loga = class_double(Loga).as_stubbed_const
+    logger = instance_double(Logger)
+    allow(loga).to receive(:logger).and_return(logger)
+    [logger, loga]
```

Comment:
It will be better if you just provide a method that inits the logger in a way that is ok for the tests. Then we can just call it at the top of the specs and do things like

```
expect(Loga.logger).to have_received(:debug).with(kind_of(Loga::Event))
```

e.g.

```
def stub_loga_logger
  Loga.reset
  Loga.configure(service_name: 'hello_world_app')
  logger = instance_double(Logger) # somehow allow all the methods so we don't have to write that in the tests https://relishapp.com/rspec/rspec-mocks/docs/basics/spies
  allow(Loga).to receive(:logger).and_return(logger)
end
```

Does it sound ok to you?

Reply:
I would rather we use verifying doubles, `allow(Loga).to receive(:logger)` won't raise an exception if `logger` isn't defined on Loga.
In addition we should be able to skip `reset` and `configure` steps for unit tests since we're only interested in what the logger receives.

We can't use a method either because we need to be able to assign the `logger` and `allow` it to receive a message.

Reply:
How about something like:

```
  def stub_logger
    loga = class_double(Loga).as_stubbed_const
    logger = instance_double(Logger)

    %i[debug info warn error fatal].each do |method|
      allow(logger).to receive(method)
    end

    allow(loga).to receive(:logger).and_return(logger)
    loga
  end

      it 'logs an info message' do
        mailer.process(event)
        expect(Loga.logger).to have_received(:debug).with(kind_of(Loga::Event))
      end
```

Reply:
I'm not convinced we should pre-emptively allow the logger to receive levels the test may be not be concerned with. I would like to see how this turns out to begin with and I would be more than happy to switch the approach you've have put forward.

Reply:
👍 Then at least make sure that you mock the `Loga.logger`:

```
def stub_logger
  loga = class_double(Loga).as_stubbed_const
  logger = instance_double(Logger)
  allow(loga).to receive(:logger).and_return(logger)
  loga
end

  
before { stub_logger }

it 'logs an info message' do
  allow(Loga.logger).to receive(:debug)
  mailer.process(event)
  expect(Loga.logger).to have_received(:debug).with(kind_of(Loga::Event))
end
```

Reply:
I can't operate on the `Loga` constant directly because it isn't mocked. We can do the following at best.

```ruby
def loga_mock
  loga = class_double(Loga).as_stubbed_const
  logger = instance_double(Logger)
  allow(loga).to receive(:logger).and_return(logger)
  loga
end


it 'logs an info message' do
  loga = loga_mock

  allow(loga.logger).to receive(:info)
  mailer.deliver(event)
  expect(loga.logger).to have_received(:info).with(kind_of(Loga::Event))
end
```

Is that what you meant?